### PR TITLE
[expo-updates][android][6/n] Create IUpdatesController and split controllers depending on configuration

### PR DESCRIPTION
**Pull request by wschurman**
# Why

There are a few disjoint systems that all were being handled by the UpdatesController:
- Invalid configuration (no URL/scope key)
- Disabled system (config disabled)
- Error initializing system (disk not readable)
- Happy path (installed, configured, can read disk, normal startup behavior, all JS API enabled)
- Dev client controller, which served two purposes: reuse code for loading updates, selectively mutates global state and reverting it when finished loading, and then behaving somewhat like a happy path afterwards.
- (Expo Go has never been handled by the UpdatesController)

The issues with all of these cases being in one controller are the following:
- implicit nullability and non-null assertions in portions of the shared code that implicitly only runs in one system/case. A good example of this is scope key nullability, which most of our system assumes in non-null, but every once in a while a path that it could be nullable in would accidentally make a non-null assertion. See https://github.com/expo/expo/pull/24330 and https://github.com/expo/expo/pull/24466.
- Same thing with isEnabled, it was being checked in a bunch of places that shouldn't even be possible to hit when not enabled.
- Same thing with emergency launch and system initialization errors: it shouldn't be possible to even create a loader if it is going to just fail down the line when it can't write to the disk.
- Dev client was mutating state, which is non-synchronized, so race conditions could occur:
    1. Dev client starts loading for the dev client load path.
    2. Some other event occurs that triggers either JS API or background load.
    3. Global state that was mutated by dev client load path now is used in this other place before it could be reverted by the dev client load path.

# How

This is fixed by making 3 separate controllers:
- UpdatesDevLauncherController - the controller that is used for dev client in all cases since dev client needs . Note that this is still pretty messy and suffers from potential race conditions, but at least it's isolated to the dev client and we can work on stability separately from the regular users. The JS API is disabled for the most part.
- DisabledUpdatesController - the controller that is used when updates is misconfigured, disabled, or can't be initialized. In this case, we can't rely upon scopeKey, runtime version, etc. It just loads the embedded update. The JS API reports that it is disabled and is mostly disabled.
- EnabledUpdatesController - the happy path controller. This is what is used for normal loading, and was the bulk of the code from the old UpdatesController, but now without all the other stuff mixed in.

# Test Plan

- Manually tested with a `et gba -n testwat expo-dev-menu expo-dev-client expo-dev-launcher expo-manifests expo-updates-interface` app to ensure dev client still works as intended. Used an app that printed Updates constants and had buttons to call the JS API methods.
- Manually disable and check disabled controller works (run test plan from https://github.com/expo/expo/pull/24466)
- CI for the happy path.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt [hunk 1]
```diff
@@ -1,55 +1,83 @@
...
+    ERROR
+  }
+
+  class Success(val update: UpdateEntity) : FetchUpdateResult(Status.SUCCESS)
+  class Failure : FetchUpdateResult(Status.FAILURE)
+  class RollBackToEmbedded : FetchUpdateResult(Status.ROLL_BACK_TO_EMBEDDED)
+  class ErrorResult(val error: Exception) : FetchUpdateResult(Status.ERROR)
+}
+
+interface IUpdatesController {
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt:59]: split this into two: public API (just start I think), internal API. make internal extend public

**Comment by wschurman** [comment]
Some e2e tests I need to add (probably in a follow-up to this and the iOS equivalent):
1. Disabled
2. dev client
3. manual setup (calling UpdatesController.initialize manually with a custom config at runtime)

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt [hunk 2]
```diff
@@ -1,55 +1,84 @@
...
```
Comment (douglowder) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt:60]: Can we put the interface in its own file?

**Review by douglowder — CHANGES_REQUESTED** [review]
See inline

**Review by douglowder — APPROVED** [review]

**Pull request merged**